modules: hal_nordic: move nrf_802154 glue to Zephyr

e-rk opened the pull request:
This commit moves all hal_nordic radio driver code that is
strictly dependent on Zephyr into the Zephyr repository.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

e-rk committed:
modules: hal_nordic: move nRF 802.15.4 Radio Driver glue to Zephyr

This commit moves all hal_nordic radio driver code that is
strictly dependent on Zephyr into the Zephyr repository.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

hubertmis reviewed (approved):
:+1:

e-rk commented on modules/hal_nordic/nrf_802154/radio/platform/nrf_802154_random_zephyr.c:
```diff
@@ -0,0 +1,61 @@
+ * and related and neighboring rights to this software to the public domain
+ * worldwide. This software is distributed without any warranty.
+ *
+ * See <http://creativecommons.org/publicdomain/zero/1.0/>.
```
The license check seems to be red due to this statement.

carlescufi replied:
The problem is that this is not a license listed by the OSI:
https://opensource.org/licenses/alphabetical

Is there any way you can avoid using code under this license?

carlescufi replied:
@nashif @MaureenHelm @galak do you know if public domain code is suitable for inclusion?

e-rk replied:
The code under the CC0 license is the PRNG. We don't use zephyr generators, because we need to get a random number from an interrupt context, possibly a zero latency interrupt.

e-rk replied:
Looking at the abovementioned file, it seems it is also under CC0.
```
/*
 * Copyright (c) 2017 Intel Corporation
 *
 * SPDX-License-Identifier: CC0-1.0
 *
 * Based on code written in 2016 by David Blackman and Sebastiano Vigna
 * (vigna@acm.org)
 *
 * To the extent possible under law, the author has dedicated all copyright
 * and related and neighboring rights to this software to the public domain
 * worldwide. This software is distributed without any warranty.
 *
 * See <http://creativecommons.org/publicdomain/zero/1.0/>.
 */
```

carlescufi replied:
Actually @e-rk you are right that we [already include](https://github.com/zephyrproject-rtos/zephyr/blob/master/subsys/random/rand32_xoroshiro128.c#L4) this very same code, so I don't see why we couldn't have another copy to be honest.

e-rk replied:
For the time being I have replaced the previous PRNG with a different one taken from OpenThread licensed under BSD 3-clause. Let me know if it's alright.

e-rk committed:
drivers: ieee802154: Fix problem with binding IRQ handlers in shim layer

The shim layer could in some circumstances not be properly
configured which would result in an unbound radio interrupt
handler.

Signed-off-by: Rafał Kuźnia <rafal.kuznia@nordicsemi.no>

rlubos commented:
The previous driver update introduced a bug, will need to wait for https://github.com/zephyrproject-rtos/zephyr/pull/32604 before this PR can be tested properly.

e-rk commented:
I have rebased the branch as there was some test failures in an unrelated sybsystem.

@nashif @MaureenHelm @galak @carlescufi 
Is it possible to get an approval to use BSD 3-clause licensed code in the file mentioned earlier?

e-rk commented:
After some discussion I have removed the BSD code completely.

tejlmand commented on modules/hal_nordic/nrf_802154/radio/CMakeLists.txt:
```diff
@@ -0,0 +1,114 @@
+
+set(NRF_802154_DRIVER_ROOT ${ZEPHYR_CURRENT_MODULE_DIR}/drivers/nrf_radio_802154)
+
+include(${NRF_802154_DRIVER_ROOT}/nrf_802154_driver_sources.cmake)
+
+zephyr_library()
+
+zephyr_include_directories(${NRF_802154_DRIVER_INCLUDE_DIRS})
+
+if (CONFIG_SOC_SERIES_NRF52X)
+zephyr_library_sources(
+  ${NRF_802154_DRIVER_SOURCES_NRF52}
+  )
+elseif (CONFIG_SOC_SERIES_NRF53X)
+zephyr_library_sources(
+  ${NRF_802154_DRIVER_SOURCES_NRF53}
+  )
+else()
+    message(FATAL_ERROR "SoC unsupported by this module")
+endif()
+
+if (CONFIG_NRF_802154_SL_OPENSOURCE)
+zephyr_library_sources(
+  ${NRF_802154_DRIVER_SOURCES_DIRECT}
+  )
+
+else()
+zephyr_library_sources(
+  ${NRF_802154_DRIVER_SOURCES_SWI}
+  )
```
This comment goes for the complete file.

I know this is existing code coming from the `hal_nordic` repo.
When CMake code is close together, creating source list or directory lists may be a sane approach.

With this PR code is moved to Zephyr, so there is an increased risk the code getting out of sync with `hal_nordic`.
As example, a user could want to use a newer `hal_nordic` with an older Zephyr or NCS release.

In such a case there could be a risk that variables are renamed between the two repos but CMake will not fail in such cases, the vars used will just be empty (undefined).
The build itself might fail, but that just makes it harder to trace back the root cause.

Or it could be a new variable introduced, as example to have SoC family specific include, like:
`NRF_802154_DRIVER_NRF53_INCLUDE_DIRS` that is not being included by a user selecting a newer revision.

Also, a minor spelling mistake like: `NRF_802145_DRIVER_INCLUDE_DIRS` will leave an empty list, but no warnings or errors in CMake.


If instead, `hal_nordic` provides a function like:
```
function(hal_nordic_sources)
  cmake_parse_arguments(HAL "DIRECT;SWI" "FAMILY;SOURCES;INCLUDES" "" ${ARGN})

  if(HAL_UNPARSED_ARGUMENTS)
    message(FATAL_ERROR "hal_nordic_sources() given unknown arguments: ${HAL_UNPARSED_ARGUMENTS}
  endif()

  if(NOT DEFINED HAL_FAMILY)
     message(FATAL_ERROR "hal_nordic_sources() missing parameter: FAMILY <NRF52|NRF53>")
  endif()

  if(DEFINED HAL_SOURCES)
    if(${HAL_FAMILY} STREQUAL NRF52)
      list(APPEND ${HAL_SOURCES} ${NRF_802154_DRIVER_SOURCES_NRF52})
    elseif(${HAL_FAMILY} STREQUAL NRF53)
      list(APPEND ${HAL_SOURCES} ${NRF_802154_DRIVER_SOURCES_NRF53})
    endif()

    if(HAL_DIRECT)
      list(APPEND ${HAL_SOURCES} ${NRF_802154_DRIVER_SOURCES_DIRECT})
    endif()

    if(HAL_SWI)
      list(APPEND ${HAL_SOURCES} ${NRF_802154_DRIVER_SOURCES_SWI})
    endif()

    set(${HAL_SOURCES} ${${HAL_SOURCES}} PARENT_SCOPE)
  endif()

  .... similar for INCLUDES .....

endfunction()
```

then Zephyr can call this function as:

```
if(CONFIG_SOC_SERIES_NRF52X)
  set(FAMILY NRF52)
elseif(CONFIG_SOC_SERIES_NRF53X)
  set(FAMILY NRF53)
endif()

if(CONFIG_NRF_802154_SL_OPENSOURCE)
  set(TYPE DIRECT)
else()
  set(TYPE SWI)
endif()


hal_nordic_sources(FAMILY ${FAMILY}
                   SOURCES  source_list
                   INCLUDES include_list
                   ${TYPE}
)

zephyr_library_sources(${source_list})

```

This gives `hal_nordic` the freedom to rename any internal variable as you might find useful.
Or in case you need to differentiate the `INCLUDES` based on soc family or something else, you may do so without require changes to Zephyr.

Introducing a new family or new options will not break anything for users not using them.

Also, if a user makes a typo, CMake will issue an error, like:
```
CMake Error at ../../modules/hal_nordic/nrf_802154/radio/CMakeLists.txt:22 (hal_nordic_sourcse):
  Unknown CMake command "hal_nordic_sourcse".
```

or if using unknown argument or unsupported family, then a proper error message is raised.

tejlmand reviewed (changes requested):
As code are now moved from `hal_nordic` into Zephyr, I would like to see a cleaner decoupling of the two repos, so that Zephyr doesn't need to know anything about internal variable in `hal_nordic`.

I have made a small proposal on how it's possible for `hal_nordic` to provide a function that makes a cleaner interface between the two.

tejlmand reviewed (approved):
As there are plans regarding better separation in the pipeline, then approving current state.

e-rk commented on modules/hal_nordic/Kconfig:
```diff
@@ -10,12 +10,19 @@ config HAS_NORDIC_DRIVERS
+config NRF_802154
+	bool
+	help
+	  This helper symbol indicates that nRF 802.15.4 radio driver API is implemented
+	  and available in some way.
```
The only reason I added this kconfig was to avoid doing the `OR` you suggested. I thought that it is more preferable to provide helper symbols rather than do some if/else logic in Cmake. I can change that back.

carlescufi reviewed (approved):
LGTM, very good that we are making progress into moving all glue code into the main tree.
